spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/DistriOptimizer.scala

```diff
@@ -786,7 +811,13 @@ class DistriOptimizer[T: ClassTag] (
     val modelsAndBroadcast = DistriOptimizer.initThreadModels(model, distDataset, criterion, state,
       nodeNumber, coresPerNode, checkSingleton, parameters, validationMethods,
       optimMethods, parameterProcessors)
-    models = modelsAndBroadcast._1
+
+    models = if (reserveOptimMethod && previousOptim != null) {
+      // replace optimMethods with previous ones
+      resetOptimMethods(modelsAndBroadcast._1, previousOptim)
```

Comment:
If we go into this branch, the current `models` will depends on `previsousOptim` which depends on the `model` on last call of `optimize()`. This may create long dependencies if we call `optimize` many times.

Reply:
Maybe we can cache the last previousOptim and release all previous ones.

Reply:
Yeah, changed. Cache the last previousOptim and release all previous ones.